src/Infrastructure.VS/ThreadHandling.cs

```diff
@@ -41,19 +40,26 @@ public class ThreadHandling : IThreadHandling
 
         public async Task RunOnUIThread(Action op) => await VS.RunOnUIThread.RunAsync(op);
 
-        public async Task<T> RunOnBackgroundThread<T>(Func<T> asyncMethod)
+        public async Task<T> RunOnBackgroundThread<T>(Func<Task<T>> asyncMethod)
         {
-            if (!ThreadHelper.CheckAccess())
-            {
-                return asyncMethod();
-            }
+            var startedOnUIThread = ThreadHelper.CheckAccess();
+
+            return await ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+                {
+                    if (startedOnUIThread)
+                    {
+                        await SwitchToBackgroundThread();
+                    }
+
+                    var result = await asyncMethod();
```

Comment:
I would just add a thread.sleep here for testing purposes, to make sure that `JoinableTaskFactory.RunAsync` doesn't block the UI thread

Reply:
I've manually tested with a ten-second sleep immediately before the execution of `asyncMethod` and the UI remains responsive.

![image](https://user-images.githubusercontent.com/10757559/182623061-c5da404e-d46b-4f8e-ac53-3fe8f064b31a.png)

This screenshot is from the slvs-community test VM. The UI thread id is 4,292.